File: src/main/java/net/softbell/bsh/controller/view/general/ReservView.java
package net.softbell.bsh.controller.view.general;

import java.util.List;

import org.springframework.security.core.Authentication;
import org.springframework.stereotype.Controller;
import org.springframework.ui.Model;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestMapping;

import lombok.AllArgsConstructor;
import net.softbell.bsh.domain.entity.NodeAction;
import net.softbell.bsh.domain.entity.NodeReserv;
import net.softbell.bsh.domain.entity.NodeReservAction;
import net.softbell.bsh.dto.request.IotReservDto;
import net.softbell.bsh.dto.view.general.ReservInfoCardDto;
import net.softbell.bsh.iot.service.v1.IotReservServiceV1;
import net.softbell.bsh.service.CenterService;
import net.softbell.bsh.service.ViewDtoConverterService;

/**
 * @Author : Bell(bell@softbell.net)
 * @Description : 예약 뷰 컨트롤러
 */
@AllArgsConstructor
@Controller
@RequestMapping("/reserv")
public class ReservView
{
	// Global Field
	private final String G_BASE_PATH = "services/general";
	private final String G_INDEX_REDIRECT_URL = "redirect:/";
	
	private final ViewDtoConverterService viewDtoConverterService;
	private final IotReservServiceV1 iotReservService;
	private final CenterService centerService;
	
	@GetMapping()
    public String dispIndex(Model model, Authentication auth)
	{
		// Exception
		if (centerService.getSetting().getIotReserv() != 1)
			return G_INDEX_REDIRECT_URL;
		
		// Field
		List<NodeReserv> listReserv;
		
		// Init
		listReserv = iotReservService.getAllReservs(auth);
		
		// Process
		model.addAttribute("listCardReservs", viewDtoConverterService.convReservSummaryCards(listReserv));
		
		// Return
        return G_BASE_PATH + "/Reserv";
    }
	
	@GetMapping("/{id}")
    public String dispReserv(Model model, Authentication auth, @PathVariable("id") long reservId)
	{
		// Exception
		if (centerService.getSetting().getIotReserv() != 1)
			return G_INDEX_REDIRECT_URL;
		
		// Field
		NodeReserv nodeReserv;
		
		// Init
		nodeReserv = iotReservService.getReserv(auth, reservId);
		
		// Process
		model.addAttribute("cardReservInfo", new ReservInfoCardDto(nodeReserv));
		model.addAttribute("listCardActionActives", viewDtoConverterService.convReservActionCards(nodeReserv.getNodeReservActions()));
		
		// Return
        return G_BASE_PATH + "/ReservInfo";
    }
	
	@GetMapping("/modify/{id}")
    public String dispReservModify(Model model, Authentication auth, @PathVariable("id") long reservId)
	{
		// Exception
		if (centerService.getSetting().getIotReserv() != 1)
			return G_INDEX_REDIRECT_URL;
		
		// Field
		NodeReserv nodeReserv;
		List<NodeAction> listNodeAction;
		
		// Init
		nodeReserv = iotReservService.getReserv(auth, reservId);
		listNodeAction = iotReservService.getAvailableAction(auth);
		
		for (NodeReservAction reservActionItem : nodeReserv.getNodeReservActions())
			listNodeAction.remove(reservActionItem.getNodeAction());
		
		// Process
		model.addAttribute("cardReservInfo", new ReservInfoCardDto(nodeReserv));
		model.addAttribute("listCardActionActives", viewDtoConverterService.convReservActionCards(nodeReserv.getNodeReservActions()));
		model.addAttribute("listCardActions", viewDtoConverterService.convReservActionCards(listNodeAction));
		
		// Return
        return G_BASE_PATH + "/ReservModify";
    }
	
	@GetMapping("/create")
    public String dispCreate(Model model, Authentication auth)
	{
		// Exception
		if (centerService.getSetting().getIotReserv() != 1)
			return G_INDEX_REDIRECT_URL;
		
		// Field
		List<NodeAction> listNodeAction;
		
		// Init
		listNodeAction = iotReservService.getAvailableAction(auth);
		
		// Process
		model.addAttribute("listCardActions", viewDtoConverterService.convReservActionCards(listNodeAction));
		
		// Return
        return G_BASE_PATH + "/ReservCreate";
    }
	
	@PostMapping("/create")
    public String procCreate(Model model, Authentication auth,
    						IotReservDto iotReservationDto)
	{
		// Exception
		if (centerService.getSetting().getIotReserv() != 1)
			return G_INDEX_REDIRECT_URL;
		
		// Field
		boolean isSuccess;
		
		// Init
		isSuccess = iotReservService.createReservation(auth, iotReservationDto);
		
		// Return
		if (isSuccess)
			return "redirect:/reserv";
		else
			return "redirect:/reserv?error";
    }
	
	@PostMapping("/modify/{id}")
    public String procModify(Model model, Authentication auth,
							@PathVariable("id") long reservId,
    						IotReservDto iotReservationDto)
	{
		// Exception
		if (centerService.getSetting().getIotReserv() != 1)
			return G_INDEX_REDIRECT_URL;
		
		// Field
		boolean isSuccess;
		
		// Init
		isSuccess = iotReservService.modifyReservation(auth, reservId, iotReservationDto);
		
		// Return
		if (isSuccess)
			return "redirect:/reserv";
		else
			return "redirect:/reserv/modify/" + reservId + "?error";
    }
	
	@PostMapping("/delete/{id}")
    public String procDelete(Model model, Authentication auth, @PathVariable("id") long reservId)
	{
		// Exception
		if (centerService.getSetting().getIotReserv() != 1)
			return G_INDEX_REDIRECT_URL;
		
		// Field
		boolean isSuccess;
		
		// Init
		isSuccess = iotReservService.deleteReserv(auth, reservId);
		
		// Return
		if (isSuccess)
			return "redirect:/reserv";
		else
			return "redirect:/reserv/modify/" + reservId + "?error";
    }
}

File: src/main/kotlin/net/softbell/bsh/controller/view/general/ReservView.kt
package net.softbell.bsh.controller.view.general

import net.softbell.bsh.domain.entity.Member
import net.softbell.bsh.domain.entity.NodeAction
import net.softbell.bsh.dto.request.IotReservDto
import net.softbell.bsh.dto.view.general.ReservInfoCardDto
import net.softbell.bsh.iot.service.v1.IotReservServiceV1
import net.softbell.bsh.service.CenterService
import net.softbell.bsh.service.ViewDtoConverterService
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.security.core.annotation.AuthenticationPrincipal
import org.springframework.stereotype.Controller
import org.springframework.ui.Model
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestMapping

/**
 * @author : Bell(bell@softbell.net)
 * @description : 예약 뷰 컨트롤러
 */
@Controller
@RequestMapping("/reserv")
class ReservView {
    // Global Field
    private val G_INDEX_REDIRECT_URL: String = "redirect:/"

    @Autowired private lateinit var viewDtoConverterService: ViewDtoConverterService
    @Autowired private lateinit var iotReservService: IotReservServiceV1
    @Autowired private lateinit var centerService: CenterService

    @GetMapping
    fun dispIndex(model: Model, @AuthenticationPrincipal member: Member): String {
        // Exception
        if (centerService.setting.iotReserv != 1.toByte())
            return G_INDEX_REDIRECT_URL

        // Init
        val listReserv = iotReservService.getPrivilegesReservs(member)

        // Process
        model.addAttribute("listCardReservs", viewDtoConverterService.convReservSummaryCards(listReserv))

        // Return
        return "services/general/Reserv"
    }

    @GetMapping("/{id}")
    fun dispReserv(model: Model, @AuthenticationPrincipal member: Member,
                   @PathVariable("id") reservId: Long): String {
        // Exception
        if (centerService.setting.iotReserv != 1.toByte())
            return G_INDEX_REDIRECT_URL

        // Init
        val nodeReserv = iotReservService.getPrivilegesReserv(member, reservId)

        // Process
        if (nodeReserv != null) {
            model.addAttribute("cardReservInfo", nodeReserv.let { ReservInfoCardDto(it) })
            model.addAttribute("listCardActionActives", viewDtoConverterService.convReservActionCards(nodeReserv.nodeReservActions))
        }

        // Return
        return "services/general/ReservInfo"
    }

    @GetMapping("/modify/{id}")
    fun dispReservModify(model: Model, @AuthenticationPrincipal member: Member,
                         @PathVariable("id") reservId: Long): String {
        // Exception
        if (centerService.setting.iotReserv != 1.toByte())
            return G_INDEX_REDIRECT_URL

        // Field

        // Init
        val nodeReserv = iotReservService.getPrivilegesReserv(member, reservId)
        val listNodeAction = iotReservService.getAvailableAction(member) as MutableList<NodeAction>
        if (nodeReserv != null) {
            for (reservActionItem in nodeReserv.nodeReservActions)
                listNodeAction.remove(reservActionItem.nodeAction)

            // Process
            model.addAttribute("cardReservInfo", ReservInfoCardDto(nodeReserv))
            model.addAttribute("listCardActionActives", viewDtoConverterService.convReservActionCards(nodeReserv.nodeReservActions))
        }
        model.addAttribute("listCardActions", viewDtoConverterService.convReservActionCards(listNodeAction))

        // Return
        return "services/general/ReservModify"
    }

    @GetMapping("/create")
    fun dispCreate(model: Model, @AuthenticationPrincipal member: Member): String {
        // Exception
        if (centerService.setting.iotReserv != 1.toByte())
            return G_INDEX_REDIRECT_URL

        // Init
        val listNodeAction = iotReservService.getAvailableAction(member)

        // Process
        model.addAttribute("listCardActions", viewDtoConverterService.convReservActionCards(listNodeAction))

        // Return
        return "services/general/ReservCreate"
    }

    @PostMapping("/create")
    fun procCreate(model: Model, @AuthenticationPrincipal member: Member,
                   iotReservationDto: IotReservDto): String {
        // Exception
        if (centerService.setting.iotReserv != 1.toByte())
            return G_INDEX_REDIRECT_URL

        // Init
        val isSuccess = iotReservService.createReservation(member, iotReservationDto)

        // Return
        return if (isSuccess)
            "redirect:/reserv"
        else
            "redirect:/reserv?error"
    }

    @PostMapping("/modify/{id}")
    fun procModify(model: Model, @AuthenticationPrincipal member: Member,
                   @PathVariable("id") reservId: Long,
                   iotReservationDto: IotReservDto): String {
        // Exception
        if (centerService.setting.iotReserv != 1.toByte())
            return G_INDEX_REDIRECT_URL

        // Init
        val isSuccess = iotReservService.modifyPrivilegesReservation(member, reservId, iotReservationDto)

        // Return
        return if (isSuccess)
            "redirect:/reserv"
        else
            "redirect:/reserv/modify/$reservId?error"
    }

    @PostMapping("/delete/{id}")
    fun procDelete(model: Model, @AuthenticationPrincipal member: Member,
                   @PathVariable("id") reservId: Long): String {
        // Exception
        if (centerService.setting.iotReserv != 1.toByte())
            return G_INDEX_REDIRECT_URL

        // Init
        val isSuccess = iotReservService.deletePrivilegesReserv(member, reservId)

        // Return
        return if (isSuccess)
            "redirect:/reserv"
        else
            "redirect:/reserv/modify/$reservId?error"
    }
}